File: 125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#

# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if str.isspace(s):
            return True
        
        left = 0
        right = len(s) - 1

        while left < right:

            if not str.isalnum(s[left]):
                left += 1
            
            if not str.isalnum(s[right]):
                right -= 1
            
            if str.isalnum(s[left]) and str.isalnum(s[right]):
                if s[left].lower() == s[right].lower():
                    left += 1
                    right -= 1
                    continue
                else:
                    return False
        return True
            

            
        
# @lc code=end

File: 125.valid-palindrome.java
/*
 * @lc app=leetcode id=125 lang=java
 *
 * [125] Valid Palindrome
 */

// @lc code=start
class Solution {
    public boolean isPalindrome(String s) {
        if (s.isEmpty()) {
            return true;
        }

        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            if (!Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }

            if (!Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }

            if (Character.isLetterOrDigit(s.charAt(left)) && Character.isLetterOrDigit(s.charAt(right))) {
                if (Character.toLowerCase(s.charAt(left)) == Character.toLowerCase(s.charAt(right))) {
                    left++;
                    right--;
                    continue;
                } else {
                    return false;
                }
            }
        }

        return true;
    }
}
// @lc code=end

